File: androidApp/app/src/test/java/com/ndefiorenze/samplemultiplatformkotlinapp/Generics.kt
package com.ndefiorenze.samplemultiplatformkotlinapp

import com.ndefiorenze.Animal
import com.ndefiorenze.AnimalBox
import com.ndefiorenze.Box
import com.ndefiorenze.Cat
import com.ndefiorenze.Dog
import com.ndefiorenze.mapIntegersToStrings
import junit.framework.Assert.assertEquals
import org.junit.Test

class Generics {

    @Test
    fun genericsBasicUsage() {
        val intBox = Box(8)
        val stringBox = Box("Hi")

        assertEquals(8, intBox.elemet)
        assertEquals("Hi", stringBox.elemet)

        // intBox.elemet = "World" <--- compilation error
    }

    @Test
    fun genericsUsage() {
        val animalBox: AnimalBox<Animal> = AnimalBox(Dog())
        assertEquals("woof", animalBox.elemet.speak())
        animalBox.elemet = Cat()
        assertEquals("meow", animalBox.elemet.speak())

        // animalBox.elemet = 2 <--- compilation error
    }

    @Test
    fun genericsMapUsage() {
        mapIntegersToStrings[1] = "one"
        mapIntegersToStrings[2] = "two"
        // mapIntegersToStrings[3] = 3 <--- compilation error
        // mapIntegersToStrings["three"] = "three" <--- compilation error
    }

}

File: iosApp/iosAppTests/Generics.swift
//
//  Generics.swift
//  iosAppTests
//
//  Created by Nicola De Fiorenze on 03/07/18.
//  Copyright © 2018 Nicola De Fiorenze. All rights reserved.
//

import XCTest
import KotlinLibrary

class Generics: XCTestCase {

    func testGenericsBasicUsage() {
        let intBox = KotlinLibraryBox(elemet: 8)
        let stringBox = KotlinLibraryBox(elemet: "Hi")
        
        XCTAssert(intBox.elemet as! Int == 8)
        XCTAssert(stringBox.elemet as! String == "Hi")
        
        intBox.elemet = "World" //<--- NO compilation error!!
        XCTAssert(intBox.elemet as! String == "World")
    }
    
    func testGenericsUsage() {
        let animalBox = KotlinLibraryAnimalBox(elemet: KotlinLibraryDog())
        XCTAssert(animalBox.elemet.speak() == "woof")
        animalBox.elemet = KotlinLibraryCat()
        XCTAssert(animalBox.elemet.speak() == "meow")
 
        //animalBox.elemet = 2 // <--- compilation error
    }
    
    func testGenericsMapUsage() {
        KotlinLibrary.mapIntegersToStrings[0] = "zero"
        KotlinLibrary.mapIntegersToStrings[1] = "one"
        KotlinLibrary.mapIntegersToStrings[2] = 2
        KotlinLibrary.mapIntegersToStrings["three"] = "three"
        
        XCTAssert("zero" == KotlinLibrary.mapIntegersToStrings[0] as! String)
        XCTAssert("one" == KotlinLibrary.mapIntegersToStrings[1] as! String)
        XCTAssert(2 == KotlinLibrary.mapIntegersToStrings[2] as! Int)
        XCTAssert("three" == KotlinLibrary.mapIntegersToStrings["three"] as! String)
    }
    
}


